src/main/java/com/gym/gymsystem/service/TrainerService.java

```diff
@@ -0,0 +1,158 @@
+    private final TrainingService trainingService;
+    private final TrainingTypeService trainingTypeService;
+
+    @Autowired
+    public TrainerService(TrainerRepository trainerRepository, UserService userService, TrainingService trainingService, TrainingTypeService trainingTypeService) {
+        this.trainerRepository = trainerRepository;
+        this.userService = userService;
+        this.trainingService = trainingService;
+        this.trainingTypeService = trainingTypeService;
+    }
+
+    @Transactional
+    public List<Trainer> getAll() {
+        return trainerRepository.findAll();
+    }
+
+    public Trainer updateTrainerProfile(UpdateTrainerProfileRequest request, Trainer trainer) {
+        if (request.getFirstName() != null) {
+            trainer.getUser().setFirstName(request.getFirstName());
+        }
+        if (request.getLastName() != null) {
+            trainer.getUser().setLastName(request.getLastName());
+        }
+        trainer.getUser().setIsActive(request.isActive());
+        return trainer;
+    }
+
+    public Trainer createTrainerProfile(Trainer trainer) {
+        logger.info("Creating trainer profile: {}", trainer);
+        userService.generateUserData(trainer.getUser());
+        List<TrainingType> trainingTypes = new ArrayList<>();
+        TrainingType trainingType = null;
+        for (TrainingType type : trainer.getSpecializations()) {
+            trainingType = trainingTypeService.getTrainingTypeByName(type.getTrainingTypeName());
+            if (trainingType == null) {
+                trainingType = new TrainingType();
+                trainingType.setTrainingTypeName(type.getTrainingTypeName());
+                trainingTypeService.save(trainingType);
+            }
+        }
+        trainingTypes.add(trainingType);
+        trainer.setSpecializations(trainingTypes);
+        return trainerRepository.save(trainer);
+    }
+
+
+    public TrainerProfileResponse updateProfile(String trainerUsername, UpdateTrainerProfileRequest request) {
+        Trainer trainer = getTrainerProfileByUsername(trainerUsername);
+        TrainingType trainingType = trainingTypeService.getTrainingTypeByName(request.getSpecialization());
+        trainer.setSpecializations(List.of(trainingType));
+        updateTrainerProfile(updateTrainerProfile(request, trainer));
+        TrainerProfileResponse response = getTrainerProfileAndTraineesByUsername(trainerUsername);
+        return response;
+    }
+
+    public Trainer getTrainerProfileByUsername(String findUsername) {
+
+        return trainerRepository.findByUser_Username(findUsername).orElseThrow(() -> new TrainerNotFoundException("Trainer not found"));
+    }
+
+    public TrainerProfileResponse getTrainerProfileAndTraineesByUsername(String findUsername) {
+        Trainer trainer = getTrainerProfileByUsername(findUsername);
+        if (trainer == null) {
+            throw new TrainerNotFoundException("Trainer not found");
+        }
```

Comment:
You are already check that trainer exist in the `getTrainerProfileByUsername` method